### PR TITLE
0.5.x Work in Progress

### PR DESCRIPTION
**Pull request by davidlehn**
**DO NOT MERGE**:  This is a work-in-progress PR tracking 0.5.x progress.

While under development the 0.5.x branch is being published to npm with the "dev" tag.  0.4.x will use the "latest" tag until 0.5.x is mature.

A main 0.5.x feature will be switching to a more modular system based on CommonJS and ES2015.  Modularity was also done in https://github.com/digitalbazaar/jsonld.js/pull/126.  That effort went fully to ES modules.  This attempt will likely start using CommonJS and have more functionality per file.  As was done in the [forge update](https://github.com/digitalbazaar/forge/milestone/1), CommonJS is used for Node.js compatibility and webpack (or similar) will be used to create browser bundles.

- [ ] Remove AMD/etc boilerplate.
- [x] Adadpt CommonJS style.
- [ ] Top level options (as in forge) for native code control.  `jsonld.options.usePureJavaScript`  Used for at least allowing native SHA256 implementation.
- [ ] webpack built browser bundles and testing.  Set up to allow building custom bundles with just needed functionality.  Aim towards future ES module design.
- [ ] Browserify built bundles and testing.
- [ ] Add docs on various tools and methods to create custom bundles
- [ ] Testing: Node.js
- [ ] Testing: In a browser manually via a webpack build of all tests
- [ ] Testing: PhantomJS, Chrome, Firefox, and Safari via Karma (webpack version)
- [ ] Testing: PhantomJS, Chrome, Firefox, and Safari via Karma (browserify version)
- [ ] Testing: Travis-CI passing Node.js and PhantomJS tests
- [ ] Docs updated
- [ ] Style and misc cleanups
- [x] Drop Bower support? It's deprecated in favor of NPM.  **Much** easier to only support built minimized bundles in NPM releases and magic NPM CDN hosting.  Alternate is to do as forge does with secondary repo and complex release process.  Make sure not to break bower usage (tags, etc).
- [ ] Update local testing.
- [ ] Improve support for local tests.
- [ ] Remove as much old build/test cruft as possible.
- [ ] Add some (not too slow) nodejs and web tests with jsonld.options.usePureJavaScript===true.
- [ ] webpack vs node/browserify test count.
- [ ] Check npm pack.
- [x] .editorconfig
- [ ] [0.5.0 Milestone Issues](https://github.com/digitalbazaar/jsonld.js/milestone/1)

As this is a rather major change it needs testing and feedback before hitting master.  Please test and leave comments, thanks!

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+// TODO: remove once Promise API is fixed
+/**
+ * Converts a node.js async op into a promise w/boxed resolved value(s).
+ *
+ * @param op the operation to convert.
+ *
+ * @return the promise.
+ */
+api.promisify = function(op) {
+  const args = Array.prototype.slice.call(arguments, 1);
```
Comment (davidlehn) [inline lib/util.js]: If this were to stay around, could do `api.promisify = function(op, ...args)` and omit the slice.

File: lib/util.js [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    return;
+  }
+  return promise.then(callback.bind(null, null), callback);
+};
+
+api.callbackify = fn => {
+  return async function() {
+    const promise = fn.apply(null, arguments);
+    if(arguments.length > fn.length) {
+      const callback = Array.prototype.slice.call(arguments).pop();
```
Comment (davidlehn) [inline lib/util.js]: Faster to do `const callback = arguments[arguments.length - 1];`

**Comment by pietercolpaert** [comment]
* [x] uses RDF.js triple representation

**Comment by adambro** [comment]
What are the plans of releasing 0.5.x version to NPM repo? I'm interested in fix for #189 issue, which has been merged to 0.5.14 apparently.

**Comment by davidlehn** [comment]
@adambro We have been releasing 0.5.x to npm.  You can use `npm install jsonld@dev` or add a dependency like `^0.5.15` if you want to live on the edge.  Early 0.5.x releases were intended to be just for development so may include breaking changes in the patch releases.  If that's a concern then you might want to use a strict version dep for now.

**Comment by davidlehn** [comment]
This branch has been merged into master and released as npm "latest".  Remaining items above will be added as issues and addressed in 0.5.x, 0.6.x, or later as appropriate.

**Pull request merged**